README.md
# Email MySQL Query Results as a CSV File Attachment
===============

A PHP class file that will email the results of a MySQL query as a CSV file attachment. 
```php
# minimum code use example
# see example.php for more options that you can use
require 'class.email-query-results-as-csv-file.php';
$emailCSV = new EmailQueryResultsAsCsv('localhost','database_name','username','password');
$emailCSV->setQuery("SELECT * FROM table_name WHERE column = 'value'");
$emailCSV->sendEmail("sender@website.com","receiver@website.com",
    "MySQL Query Results as CSV Attachment");
```

This can be run by opening it in a web browser, or run as a CRON job. The more common way to use it, is to integrate it into some other program. The source code in this PHP file is well commented and can be modified to fit your needs.

## License
This software is licensed under the [GNU GPL version 3.0](http://www.gnu.org/licenses/gpl-3.0-standalone.html) or later.

## What it does do:

1. Connects to a MySQL Database.
2. Runs one (or more) MySQL query that you define.
3. Builds a correctly formatted CSV file from the query results. Can send one or many CSV file attachements in a single email.
4. Emails the CSV file as an attachment to an address you specify. 

## What it doesn't do:

This code doesn't save the results to a file on the server.
If you are looking to save a file to the server, try Google search.

# Basic Trouble Shooting

## Problems Sending The Email?
  
If you receive this message "ERROR: The Email could not be sent." 
  
Then you probably just need to configure your mail server. The situation where I see this the most is with people running WAMP as a local development server. Check with your web host to find out your mail server settings. You will need to know:
* the address of the mail server
* the port to use on the mail server
* the address to use that you will send mail from

Then lookup the PHP function [ini_set()](http://php.net/manual/en/function.ini-set.php)

Finally add to the top of this file code like this...
```php
ini_set("SMTP","mail.domain.com");
ini_set("SMTP_PORT", 25);
ini_set("sendmail_from","yourname@domain.com");
```

## The script says the mail was sent, but it has never been received.

### Script Says Email Sent

The PHP mail() function returns true when the outgoing mail server accepts the message for delivery. This doesn't mean the email has been delivered correctly. 

### Example Successful Script Output
```php
Step 1: Connected to MySQL server successfully. Step 2: MySQL database successfully 
selected. Step 3 (repeats for each attachment): MySQL query ran successfully. Step 4 
(repeats for each attachment): CSV file built. Step 5: Email sent with attachment. 
FINISHED.
```

### Some Things To Check

There are many issues that could cause email sent via the PHP mail() function to not be received. 

You will need to troubleshoot possibilities, unrelated to this script, to find the point of failure. Some examples:
* The mail was received, but marked as spam.
* The recipient's address was incorrect.
* There is a slow mail queue on the sending mail server.
* There is a slow mail queue on the receiving mail server.

The problem is likely a server configuration problem, you will want to contact your web hosting provider.

### Test Sending An Email Using PHP mail() Function
Make sure you can still send email from your server. Once you know if you can send email from your server, you can move forward with other possible solutions. You can also try sending the mail to different recipients to help narrow down the problem. 

If you have ssh/command line access, you can try this:

```
php -r 'mail( 'you@yourdomain.com', "subject: test", "message: test", 'From: you@yourdomain.com');'
```

If you have FTP access, put this into a file called mailer.php (or whatever.php you want) and then navigate to it with a browser:

```php
<?php
  mail( 'you@yourdomain.com', "subject: test", "message: test", 'From: you@yourdomain.com'); 
  echo "Mail should have been sent, check your inbox"; 
?>
```

# New beginner instructions on how to use the class, and example files.

## Step 1: View then Edit the Example File

The source code is well documented. Start by opening the driver file "example.php" in your code editor. You could try Notepad++ for windows, and gEdit (pimped) for Linux.

* Each of the lines of code is very well commented and documented in this file.
* Follow the directions outlined in comments of the file.
* Some items are optional, and they are noted as such.

## Step 2: Upload the Class File and Driver File to Your Web Sever

After you fill in your specific details, upload the two files to your web server:

* class.email-query-results-as-csv-file.php
* example.php

## Setp 3: Run the Driver File from a Web Browser

Run the file "example.php" from a web browser.

## debugMode(True)

If you are using the object in debugMode(True) then it will output messages of success or error at each step. Error messages are always output regardless if the debugMode is active or not.

## Attaching Multiple Files to One Email

To attach multiple files to a single email follow the code in the file "example-multi.php".

# Best of luck to you.
Ask questions in the comments on this [project's code page at Studio-Owens.com](http://www.studio-owens.com/code/email-mysql-query-results-as-a-csv-file-attachment.htm)


history.txt
Email MySQL Query Results as a CSV File Attachment
file: class.email-query-results-as-csv-file.php

Author:  Stephen R. Owens (www.Studio-Owens.com)

Sends an email with a CSV file attachment that contains the results of a MySQL query.
Copyright (C) 2009-2014 Stephen R. Owens


HISTORY:

2014-MAR-22 : v2.2
* Moved repo from Google Code to GitHub

2014-FEB-21 : v2.2
* Removed support for the deprecated MySQL extension
+ added MySQLi support

2014-FEB-20 : v2.1.3
* Switched repo from SVN to Git - no code changes

2013-DEC-11 : v2.1.3
* Changed the character set from iso-8859-1 to UTF-8 for better international character support within the body of the generated email

2012-FEB-21 : v2.1.2
* bug fix: num of rows returned by the query must be greater than zero; previous check caused no file to be built in v2.1.1

2012-FEB-12 : v2.1.1
+ added behaviour to only attach file when at least one row has been returned from the database
+ added if no files were attached do not send email

2011-MAR-07 : v2.1
+ added the ability to attach multiple files
+ new file "example-multi.php" shows how to use the class to send multiple file attachments
+ moved setting the CSV file name to the method "setCSVname()"
* the method "setCSVinfo()" is now used to only set the reusable parts of the CSV file, and no longer sets the file name
* changed "example.php" and "example-min.php" to show the new use of setCSVname()

2011-MAR-06 : v2.0.1
- Removed the extra comma at the end of each row

2011-MAR-02 : v2.0
* Changed the code to a class file, because (a) it's better to use OOP and (b) encourage people to leave the copyright in place as is required by the GNU GPL license.
+ new file: "example.php" to show how the class is used with comments explaining the code
+ new file "example-min.php" to show the minimum amount of code needed to use this class

2009-NOV-11 : v1.0
* Initial Public Release


example-multi.php
<?php
/*
  This is an example of attaching multiple files to a single email. After the email is sent the file
  attachment list is cleared so this same object could be used again for a new email.
*/

# include the class and create a new object of the class
require 'class.email-query-results-as-csv-file.php';
$emailCSV = new EmailQueryResultsAsCsv('localhost','database_name','username','password');

# -----
# Attach Multiple Files To One Email
# you can attach multiple files by setting them up using the object method "setMultiFile()"
# pass in this information: (a) File Name for this attachment, (b) Query
# repeat as needed, method can be called once or multiple times before sending the email
$emailCSV->setMultiFile("file_one.csv", "SELECT * FROM table_name_one");
$emailCSV->setMultiFile("file_two.csv", "SELECT * FROM table_name_two");
$emailCSV->setMultiFile("file_three.csv", "SELECT * FROM table_name_three");

# send the email
$emailCSV->sendEmail("sender@website.com","receiver@website.com","MySQL Query Results as CSV Attachment");
?>


class.email-query-results-as-csv-file.php
<?php
/*
Email MySQL Query Results as a CSV File Attachment
Author:  Stephen R. Owens  (www.Studio-Owens.com)
Version: 2.2 [2:41 AM Saturday, February 22, 2014]

Sends an email with a CSV file attachment that contains the results of a MySQL query.
Copyright (C) 2009-2014 Stephen R. Owens

LICENSE:
This software is licensed under the GNU GPL version 3.0 or later.
http://www.gnu.org/licenses/gpl-3.0-standalone.html

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

class EmailQueryResultsAsCsv {

  # MySQL Server
  private $mySQL_server = '';
  private $mySQL_server_port = '3306';

  # MySQL Database Name
  private $mySQL_database = '';

  # MySQL Username
  private $mySQL_user = '';

  # MySQL Password
  private $mySQL_password = '';

  # MySQL Query
  # something like "SELECT * FROM table_name"
  private $mySQL_query = '';

  # CSV File Name
  # filename for the attached file, use something like "mysql_results.csv"
  private $csv_file_name = '';
  
  # Multiple File Data array
  # File Name & reuseable elements + mySQL_query
  private $arr_file_data = array();
  
  # CSV file reuseable elements
  private $csv_contain = '';
  private $csv_separate = '';
  private $csv_end_row = '';

  # Email Message
  # This is an HTML formatted message
  private $email_html_msg = "<h1>MySQL Query Results as CSV Attachment</h1>
  <p>This attachment can be opened with OpenOffice.org Calc, Google Docs, or Microsoft Excel.</p>";
  
  # used to output success messages to the screen
  private $debugFlag = False;
  
  # -------------------------------------- 
  #  Methods
  # -------------------------------------- 
    
  # constructor
  public function __construct($s, $d, $u, $p) {
    $this->setDBinfo($s, $d, $u, $p);
    $this->setCSVinfo();
    $this->setCSVname();
  }
  
  # destructor
  public function __destruct() {
    
  }
  
  public function setDBinfo($s, $d, $u, $p) {
    $this->mySQL_server = $s;
    $this->mySQL_database = $d;
    $this->mySQL_user = $u;
    $this->mySQL_password = $p;
  }
  
  public function setDBinfoServerPort($p) {
    $this->mySQL_server_port = $p;
  }
  
  public function setQuery($sql) {
    $this->mySQL_query = $sql;
  }
  
  public function setEmailMessage($msg) {
    $this->email_html_msg = $msg;
  }
  
  public function setCSVname($fn = "mysql_results.csv") {
    $this->csv_file_name = $fn;
  }
  
  public function setCSVinfo($c = '"', $s = ",", $er = "\n") {
    $this->csv_contain = $c;
    $this->csv_separate = $s;
    $this->csv_end_row = $er;
  }
  
  public function setMultiFile($fn, $sql) {
    $this->arr_file_data[] = array("csv_file_name" => $fn, 
      "mySQL_query" => $sql,
      "csv_contain" => $this->csv_contain,
      "csv_separate" => $this->csv_separate,
      "csv_end_row" => $this->csv_end_row);
  }
  
  public function debugMode($bool) {
    $this->debugFlag = $bool;
  }
  
  public function sendEmail($email_from, $email_to, $email_subject) {
    # check to see if the array for file info and queries has data if not add the single file data
    if(!isset($this->arr_file_data[0]["csv_file_name"])) {
      $this->arr_file_data[0] = array("csv_file_name" => $this->csv_file_name, 
        "mySQL_query" => $this->mySQL_query,
        "csv_contain" => $this->csv_contain,
        "csv_separate" => $this->csv_separate,
        "csv_end_row" => $this->csv_end_row);
    }

    # -------------------------------------- 
    #   CONNECT TO MYSQL DATABASE
    $mysqli = new mysqli($this->mySQL_server, $this->mySQL_user, $this->mySQL_password, $this->mySQL_database, $this->mySQL_server_port);
    if ($mysqli->connect_errno) {
         die('ERROR: Could not connect to MySQL server: (' . $mysqli->connect_errno . ') ' . $mysqli->connect_error);
    }
    if ($this->debugFlag) {
      echo "Step 1: Connected to MySQL server successfully. \n\n";
    }
    
    /*
    -------------------------------------- 
        SEND EMAIL WITH ATTACHMENT
        Below you will notice I'm using complex variable parsing {curly braces surrounding variable}
          For more information see: "PHP: Complex Variables in Strings" at Tinsology.net
          http://tinsology.net/2009/06/php-complex-variables-in-strings/
    */

    # start setting up the email header
    $headers = "From: ".$email_from;

    # create boundary string
    # boundary string must be unique using MD5 to generate a pseudo random hash
    $random_hash = md5(date('r', time())); 
    $mime_boundary = "==Multipart_Boundary_x{$random_hash}x";

    # set email header as a multipart/mixed message
    # this allows the sending of an attachment combined with the HTML message
    $headers .= "\nMIME-Version: 1.0\n" .
    "Content-Type: multipart/mixed;\n" .
    " boundary=\"{$mime_boundary}\"";

    # multipart boundary for the HTML message
    $email_message = "This is a multi-part message in MIME format.\n\n" .
    "--{$mime_boundary}\n" .
    "Content-Type:text/html; charset=\"UTF-8\"\n" .
    "Content-Transfer-Encoding: 7bit\n\n" .
    $this->email_html_msg . "\n\n";
    
    # count how many files were created
    $files_attached_cnt = 0;
    foreach ($this->arr_file_data AS $file_data) {
      
      # build the CSV file
      $csv_file = $this->buildCSV($file_data, $mysqli);
      
      if ($csv_file != "") {
        $files_attached_cnt += 1;
      
        # encode CSV file with MIME base64
        # required for sending it as an email attachment
        $data = chunk_split(base64_encode($csv_file)); 
        
        # multipart boundary for the email attachment
        $email_message .= "--{$mime_boundary}\n";
        
        # attach the file
        $email_message .= "Content-Type: application/octet-stream;\n" .
        " name=\"{$file_data['csv_file_name']}\"\n" .
        "Content-Disposition: attachment;\n" .
        " filename=\"{$file_data['csv_file_name']}\"\n" .
        "Content-Transfer-Encoding: base64\n\n" .
        $data . "\n\n";
      }
    }
    # end the multipart message with this mime boundary, notice the ending --
    $email_message .= "--{$mime_boundary}--\n";
    
    if ($files_attached_cnt > 0) {
      # try to send the email and verify the results

      $sendit = @mail($email_to, $email_subject, $email_message, $headers);
      if(!$sendit) {
        die("ERROR: The Email could not be sent.");
      }

      if ($this->debugFlag) {
        echo "Step 4: Email sent with attachment. \n\n";
      }
    } else {
      if ($this->debugFlag) {
        echo "Step 4: No data found for query. Email NOT sent. \n\n";
      }
    }

    # close the link to the MySQL database
    //$mysqli->close();
    
    # reset the attachment array so the object can be used anew
    $this->arr_file_data = array();

    if ($this->debugFlag) {
      echo "FINISHED.";
    }
  }
  
  private function buildCSV($file_data, $mysqli) {
    # container to hold the CSV file as it's built
    $csv_file = "";
    
    # run the MySQL query and check to see if results were returned
    $result = $mysqli->query($file_data["mySQL_query"]);
    if (!$result) {
      die("ERROR: Invalid query \n MySQL error: (" . $mysqli->errno . ")" . $mysqli->error . "\n Your query: " . $this->mySQL_query);
    }
    
    # only return a non blank data set with query returns at least one record
    if ($result->num_rows > 0) {
      if ($this->debugFlag) {
        echo "Step 2 (repeats for each attachment): MySQL query ran successfully. \n\n";
      }
      
      # store the number of columns and field data from the results
      $columns = $mysqli->field_count;
      $column_data = $result->fetch_fields();
      
      # Build a header row using the mysql field names
      $header_row = '';
      $i = 0;
      foreach ($column_data as $col) {
      //for ($i = 0; $i < $columns; $i++) {
        $column_title = $file_data["csv_contain"] . stripslashes($col->name) . $file_data["csv_contain"];
        $column_title .= ($i < $columns-1) ? $file_data["csv_separate"] : ''; #the last column does not have the column separator
        $header_row .= $column_title;
        $i++;
      }
      $csv_file .= $header_row . $file_data["csv_end_row"]; # add header row to CSV file
      
      # Build the data rows by walking through the results array one row at a time
      $data_rows = '';
      while ($row = $result->fetch_array(MYSQLI_NUM)) {
        for ($i = 0; $i < $columns; $i++) {
          # clean up the data; strip slashes; replace double quotes with two single quotes
          $data_rows .= $file_data["csv_contain"] . preg_replace('/'.$file_data["csv_contain"].'/', $file_data["csv_contain"].$file_data["csv_contain"], stripslashes($row[$i])) . $file_data["csv_contain"];
          $data_rows .= ($i < $columns-1) ? $file_data["csv_separate"] : '';
        }
        $data_rows .= $this->csv_end_row; # add data row to CSV file
      }
      $csv_file .= $data_rows; # add the data rows to CSV file
      
      if ($this->debugFlag) {
        echo "Step 3 (repeats for each attachment): CSV file built. \n\n";
      }
    }  else {
       echo "Step 2 (repeats for each attachment): MySQL query ran successfully \n\n";
       echo "Step 3 (repeats for each attachment): NO results were returned for this query. No file will be sent for the following query: \n " . $this->mySQL_query ." \n\n";
    }
    
    # Return the completed file
    return $csv_file;
  }
}
?>

example.php
<?php
/*
  Modify this example as you see fit. 
  This file could be run as a CRON JOB or with a direct file call.
*/

# -----
# Load The Class File
# first we have to load the class file
require 'class.email-query-results-as-csv-file.php';

# -----
# Create A New Object
# next we create a new object of the class "EmailQueryResultsAsCsv"
# we must pass in the following information upon creating the new object: 
# (a) MySQL Server, (b) MySQL Database Name, (c) MySQL Username, (d) MySQL Password
$emailCSV = new EmailQueryResultsAsCsv('localhost','database_name','username','password');

# -----
# [Optional] 
# Set the database server port - default is the standard 3306
$emailCSV->setDBinfoServerPort('3306');

# -----
# [Optional] 
# DEBUG outputs text to the screen
# you can set this to False or just delete the line all together to keep success messages
# from being printed to the screen; note that ERROR messages will always be shown if one occurs
$emailCSV->debugMode(True);

# -----
# [Optional] 
# CSV File Name
# default is "mysql_results.csv" but you can set it to be something more descriptive
$emailCSV->setCSVname("mysql_results.csv");

# -----
# Database Query
# set the query that you want to run that will be used to fill the CSV file
# example "SELECT * FROM table_name WHERE column = 'value'"
$emailCSV->setQuery("SELECT * FROM table_name");

# -----
# [Optional] 
# Email Message HTML Formatted 
# supply a message to be sent in the body of the email
$emailCSV->setEmailMessage("<h1>MySQL Query Results as CSV Attachment</h1><p>This attachment can be opened with OpenOffice.org Calc, Google Docs, or Microsoft Excel.</p>");

# -----
# [Optional] 
# Set the reusable parts of the CSV file, default options are shown in the method call below
# Excel, OpenOffice.org Calc, and Google Spreadsheet will all use the default settings just fine
# example use would to format the CSV file to the German CSV standard instead of the USA/UK format
# Fields with embedded data enclosure characters will be enclosed within a double-data enclosure, 
#   e.g. default setting action is that each embedded double-quote character must be represented by
#   a pair of double-quote characters
#   more info on the CSV file format: http://en.wikipedia.org/wiki/Comma-separated_values
# pass in the following variables: 
#   (a) data enclosure, (b) value separator, (c) newline
$emailCSV->setCSVinfo('"', ",", "\n");

# -----
# Send The Email With CSV File Attachment
# this runs the query, builds the CSV file, and attaches it to an email
# to send the email you must pass in the following information:
# (a) email sender, (b) email receiver, (c) subject of the email 
$emailCSV->sendEmail("sender@website.com","receiver@website.com","MySQL Query Results as CSV Attachment");

/*****************************************************************
  Problems Sending The Email?
  
  If you receive this message "ERROR: The Email could not be sent." 
  
  Then you probably just need to configure your mail server. The 
  situation where I see this the most is with people running WAMP 
  as a local development server. Check with your web host to find 
  out your mail server settings. You will need to know:
    * the address of the mail server
    * the port to use on the mail server
    * the address to use that you will send mail from
  Then lookup the PHP function ini_set()
    http://php.net/manual/en/function.ini-set.php
  Finally add to the top of this file code like this...
    ini_set("SMTP","mail.domain.com");
    ini_set("SMTP_PORT", 25);
    ini_set("sendmail_from","yourname@domain.com");
*****************************************************************/
?>
